### PR TITLE
feat: add contributor insights to workspaces

### LINKED ISSUE #2652: Feature: List pages should be under a workspaces route
As a user of workspaces, I want all my current list pages to be accessible via a workspace page route, i.e. `/workspaces/some-id/contributor-insights/some-id`

### PR DESCRIPTION
**Pull request by brandonroberts**
## Description

- Adds navigation from workspaces to contributor insights through workspace URLs
- Most of the additions are from duplicating the current hub/* pages into contributor insight pages and updating the links to navigate within workspaces
- Adds support for workspace URLs to shared components

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Closes #2652

## Mobile & Desktop Screenshots/Recordings

 

https://github.com/open-sauced/app/assets/42211/8e1a2fd4-89a9-4999-a0bb-d396fd9b595a

## Steps to QA

1. Login
2. Navigate to https://deploy-preview-2648--oss-insights.netlify.app/workspaces/new
3. Click on Contributor Insights on the side nav
4. Click on Contributor Insight from the list
5. Visit tabs
6. Create a new contributor insight

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: components/AddContributorsHeader/add-contributors-header.tsx [hunk 1]
```diff
@@ -13,6 +13,7 @@ import useDebounceTerm from "lib/hooks/useDebounceTerm";
 interface AddContributorsHeaderProps {
   selectedContributorsIds: number[];
   list: DbUserList;
+  workspaceId?: string;
```
Comment (nickytonline) [inline components/AddContributorsHeader/add-contributors-header.tsx:16]: The places where `workspaceId` is optional is temporary until we flip the switch to enable workspaces for everyone?

File: components/molecules/ListCard/list-card.tsx [hunk 2]
```diff
@@ -26,7 +27,15 @@ const ListCard = ({ list, handleOnDeleteClick }: ListCardProps) => {
           <div className="flex items-center gap-4 lg:items-center ">
             <div className="w-4 h-4 rounded-full bg-light-orange-10"></div>
             <div className="flex justify-between text-xl text-light-slate-12">
-              <Link href={`/lists/${list.id}/overview`}>{list.name}</Link>
+              <Link
+                href={
+                  workspaceId
+                    ? `/workspaces/${workspaceId}/contributor-insights/${list.id}/overview`
+                    : `/lists/${list.id}/overview`
```
Comment (nickytonline) [inline components/molecules/ListCard/list-card.tsx:34]: The old URLs are there temporarily until we're all in on workspaces?

File: pages/workspaces/[workspaceId]/contributor-insights/[listId]/highlights.tsx [hunk 3]
```diff
@@ -0,0 +1,272 @@
+import React, { useEffect, useMemo, useRef, useState } from "react";
```
Comment (nickytonline) [inline pages/workspaces/[workspaceId]/contributor-insights/[listId]/highlights.tsx:1]: I imagine the list pages under workspaces are literal copy paste from the the old list routes?

**Review by nickytonline — APPROVED** [review]
This is working great! Left a couple comments, but this is good to go.

File: components/molecules/ListCard/list-card.tsx (hunk 2 above)
Comment (brandonroberts) [reply components/molecules/ListCard/list-card.tsx:34]: Yes, that's correct. This is used on the old and new pages

File: components/AddContributorsHeader/add-contributors-header.tsx (hunk 1 above)
Comment (brandonroberts) [reply components/AddContributorsHeader/add-contributors-header.tsx:16]: Yep, these are used in both lists and workspaces for now

File: pages/workspaces/[workspaceId]/contributor-insights/[listId]/highlights.tsx (hunk 3 above)
Comment (brandonroberts) [reply pages/workspaces/[workspaceId]/contributor-insights/[listId]/highlights.tsx:1]: Yes, copy/paste and updated links for navigation from within workspaces

**Comment by nickytonline** [comment]
The only thing to note here is that all existing lists will be associated to a personal workspace once @jpmcb work is done. This should be fine for the redirect rule you have in place, but just wondering if there's a scenario post personal workspaces that you might not be on your personal workspace and someone hits the old link to get redirected to a route where the workspace ID isn't associated to the list.

I could be overthinking this. 😅

**Comment by brandonroberts** [comment]
> The only thing to note here is that all existing lists will be associated to a personal workspace once @jpmcb work is done. This should be fine for the redirect rule you have in place, but just wondering if there's a scenario post personal workspaces that you might not be on your personal workspace and someone hits the old link to get redirected to a route where the workspace ID isn't associated to the list.
> 
> I could be overthinking this. 😅

😅  I've thought about this also. I was thinking we can find the workspace ID through the API for the "owner" of contributor insight and redirect them server-side

File: components/AddContributorsHeader/add-contributors-header.tsx [hunk 4]
```diff
@@ -42,7 +44,9 @@ const AddContributorsHeader = ({
         <h1 className="self-start text-2xl flex items-center">
           <Link
             className="inline-block p-3 mr-2 border rounded-lg cursor-pointer bg-light-slate-1"
-            href={`/lists/${list.id}/edit`}
+            href={
+              workspaceId ? `/workspaces/${workspaceId}/contributor-insights/${list.id}/edit` : `/lists/${list.id}/edit`
+            }
```
Comment (jpmcb) [inline components/AddContributorsHeader/add-contributors-header.tsx:49]: This redirect should work 👍🏼 with personal workspaces/other workspaces: in the end, they're all just workspaces. More details here: https://github.com/open-sauced/api/pull/558

**Review by jpmcb — APPROVED** [review]

**Pull request merged**